sklearn/_loss/loss.py

```diff
@@ -461,6 +461,57 @@ def constant_to_optimal_zero(self, y_true, sample_weight=None):
         """
         return np.zeros_like(y_true)
 
+    def init_gradient_and_hessian(self, n_samples, dtype=np.float64, order="F"):
```

Comment:
Yup, this looks good.

Reply:
> (Interestingly, the gradient_out=None does not allow dtype and order to be specified.)

Would you expect such arguments for the case `gradient_out=None`?
I think, in the end, all estimators might want to initialize gradients themselves and then reuse it. Until then, `gradient_out=None` makes life easier.

Reply:
Looking a bit deeper: If `gradient_out=None`, an array is created with `gradient_out = np.empty_like(raw_prediction)`. I think that's good enough.

Reply:
> Would you expect such arguments for the case gradient_out=None?

No, I think it would be too much.

With `gradient_out=None`, `np.empty` defaults to `order="C"` which is slightly different from `init_gradient_and_hessian`. I think we should at least match `np.empty(dtype=)` with `init_gradient_and_hessian(dtype=)`.

Reply:
I don't understand what you suggest. Current behaviour:
- In the loss, gradient and hessian functions, if, e.g., `gradient_out=None` then `gradient_out = np.empty_like(raw_prediction)` is called which uses shape, dtype and order of `raw_prediction`.
- `init_gradient_and_hessian(.., dtype=.., order=..)` uses the arguments `dtype` and `order` when calling `np.empty(shape=shape, dtype=dtype, order=order)`. There is nothing else from which to infer those parameters.
- `predict_proba` of `HalfBinomialLoss` is maybe the only inconsistent case. It creates C-contiguous arrays instead of same order as `raw_prediction`.

Reply:
Ah sorry, I misread the current implementation. The defaults for `init_gradient_and_hessian` make sense to me now.